go/types/type.go

```diff
@@ -111,41 +111,33 @@ func TypeOf(v Value) *Type {
 
 // HasStructCycles determines if the type contains any struct cycles.
 func HasStructCycles(t *Type) bool {
-	return hasStructCycles(t, map[string]int{})
+	return hasStructCycles(t, nil)
 }
 
-func hasStructCycles(t *Type, seenStructs map[string]int) bool {
+func hasStructCycles(t *Type, visited []*Type) bool {
```

Comment:
Not sure this is better? It comes down to higher constant factor O(1) vs lower constant factor O(n).

But lets keep the slice solution. It is simpler to understand.